### PR TITLE
feat(gatsby-source-drupal): Persist backreference data individually in cache so cheaper to get/set

### PR DESCRIPTION
**Pull request by KyleAMathews**
Previously we get/set a big object of all back references on every sourcing. This gets slow for big Drupal sites so this PR switches to getting/setting individual records which is much cheaper as they're rarely updated after first creating.

### CODE REVIEW

File: packages/gatsby-source-drupal/src/gatsby-node.js [hunk 1]
```diff
@@ -659,14 +653,17 @@ ${JSON.stringify(webhookBody, null, 4)}`
...
   // second pass - handle relationships and back references
-  nodes.forEach(node => {
-    handleReferences(node, {
-      getNode: nodes.get.bind(nodes),
-      mutateNode: true,
-      createNodeId,
-      entityReferenceRevisions,
+  await Promise.all(
+    Array.from(nodes.values()).map(node => {
+      handleReferences(node, {
```
Comment (TylerBarnes) [inline packages/gatsby-source-drupal/src/gatsby-node.js:658]: I think you need to return the handleReferences promise here if you want to Promise.all it

File: packages/gatsby-source-drupal/src/utils.js [hunk 2]
```diff
@@ -113,35 +94,39 @@ const handleReferences = (
     })
 
     delete node.drupal_relationships
-    referencedNodesLookup.set(node.id, referencedNodes)
+    cache.set(makeRefNodesKey(node.id), referencedNodes)
```
Comment (TylerBarnes) [inline packages/gatsby-source-drupal/src/utils.js:97]:
```suggestion
    await cache.set(makeRefNodesKey(node.id), referencedNodes)
```
Do we need to await here as well?

**Review by TylerBarnes — COMMENTED** [review]
Nice 🙌 Just a couple comments

File: packages/gatsby-source-drupal/src/gatsby-node.js (hunk 1 above)
Comment (KyleAMathews) [reply packages/gatsby-source-drupal/src/gatsby-node.js:658]: 🤦

File: packages/gatsby-source-drupal/src/utils.js [hunk 3]
```diff
@@ -181,22 +167,22 @@ const handleDeletedNode = async ({
...
-  referencedNodesLookup.delete(deletedNode.id)
+  cache.del(makeBackRefsKey(deletedNode.id))
+  cache.del(makeRefNodesKey(deletedNode.id))
 
   // Remove relationships from other nodes and re-create them.
   Object.keys(deletedNode.relationships).forEach(key => {
     let ids = deletedNode.relationships[key]
     ids = [].concat(ids)
-    ids.forEach(id => {
+    ids.forEach(async id => {
```
[the remark this answers is not in the document]
Comment (KyleAMathews) [reply packages/gatsby-source-drupal/src/utils.js:177]: yes

File: packages/gatsby-source-drupal/src/utils.js [hunk 4]
```diff
@@ -113,35 +94,39 @@ const handleReferences = (
...
+          let backRefsNames = await cache.get(
+            makeBackRefsKey(referencedNode.id)
+          )
+          if (!backRefsNames) {
+            backRefsNames = []
+            cache.set(makeBackRefsKey(referencedNode.id), backRefsNames)
+          }
+
+          if (!backRefsNames.includes(nodeFieldName)) {
+            backRefsNames.push(nodeFieldName)
```
[the remark this answers is not in the document]
Comment (KyleAMathews) [reply packages/gatsby-source-drupal/src/utils.js:125]: yes, fixed

**Pull request merged**